src/app/core/components/drag-n-drop/drag-n-drop.component.ts

```diff
@@ -0,0 +1,213 @@
+    this.store.select(getFormItems).pipe(
+      takeUntil(this.ngUndestroy$)
+    ).subscribe(
+      (elements) => {
+        this.buildForm(elements);
+        this.formElements = [ ...elements ];
+        elements.forEach(e => {
+          this.setValidatorsToForm(e);
+        });
+      }
+    );
+  }
+
+  private buildForm(controls: NameValueInterface[]): void {
+    controls.forEach(el => {
+      switch (el.name) {
+        case EElementNames.input:
+          this.form.addControl(el.id, new FormControl(''));
+          break;
+        case EElementNames.checkbox:
+          this.form.addControl(el.id, new FormControl(false));
+          break;
+        case EElementNames.textarea:
+          this.form.addControl(el.id, new FormControl(''));
+          break;
+        case EElementNames.select:
+          this.form.addControl(el.id, new FormControl(''));
+          break;
+      }
+    });
+  }
+
+  private setValidatorsToForm(item: NameValueInterface): void {
+    if (item.name !== EElementNames.button) {
+      if (item.required) {
+        this.form.get(item.id).setValidators(item.name !== 'checkbox' ? [ Validators.required ] : [ Validators.requiredTrue ]);
+      } else {
+        this.form.get(item.id).clearValidators();
+      }
+    }
+  }
+
+  drop(event: CdkDragDrop<object[]>): void {
+    this.initElements = [ ...this.constInitElements ];
+    if (event.previousContainer === event.container) {
+      moveItemInArray(event.container.data, event.previousIndex, event.currentIndex);
+    } else {
+      transferArrayItem(event.previousContainer.data,
+        event.container.data,
+        event.previousIndex,
+        event.currentIndex);
+      this.formElements[event.currentIndex].id = `${this.formElements[event.currentIndex].name}-${Date.now()}`;
+    }
+    this.store.dispatch(actions.updateFormItem({ payload: JSON.parse(JSON.stringify(this.formElements)) }));
+    this.cdr.detectChanges();
+  }
+
+  noReturnPredicate(): boolean {
+    return false;
+  }
+
+  delElement(id: string, index: number): void {
+    this.formElements.splice(index, 1);
+    Object.keys(this.form.value).forEach(
+      key => key !== 'checkbox'
+        ? this.form.get(key).setValue('')
+        : this.form.get(key).setValue(false)
+    );
+    this.form.removeControl(id);
+    this.store.dispatch(actions.updateFormItem({ payload: JSON.parse(JSON.stringify(this.formElements)) }));
+  }
+
+  saveForm(gStyles: NameValueInterface): void {
+    if (this.builder) {
+      this.builder.builderArray = this.formElements;
+      this.builder.generalStyles = gStyles;
+      this.formBuilderService.saveFormBuilder(this.builder).subscribe();
+    } else {
+      const builder: FormBuilderModel = {
+        userId: this.getFormBuilderId(),
+        builderArray: [ ...this.formElements ],
+        generalStyles: this.generalStylesElement
+      };
+      this.formBuilderService.addFormBuilder(builder).subscribe();
+    }
+  }
```

Comment:
seems like code duplication, find a better solution

Reply:
move to service and optimize